remove read_excel keyword NotImplemented error, update documentation #11544

grahamjeffries opened the pull request:
Followed up in #14326

---

Towards fixing bug #11544, with respect to comments on merged PR #11870.

I've removed the `NotImplemented` error for `parse_dates` and `date_parser` keywords in `read_excel`, seeing that they are implemented (see discussion in #11870). I've added an example of the intended functionality in the documentation

grahamjeffries committed:
remove read_excel kwd NotImplemented error, update documentation #11544

jorisvandenbossche commented on doc/source/io.rst:
```diff
@@ -2192,6 +2192,18 @@ indices to be parsed.
++++++++++++++
+
+The `parse_dates` keyword for `read_excel` is used to specify whether to parse strings
+to a datetime.
```
Maybe note specifically that you do *no* need to use this when the data are formatted as dates in the excel file.

jorisvandenbossche commented on pandas/io/excel.py:
```diff
@@ -298,13 +298,10 @@ def _parse_excel(self, sheetname=0, header=0, skiprows=None, skip_footer=0,
-        if date_parser is not None:
-            raise NotImplementedError("date_parser keyword of read_excel "
-                                      "is not implemented")
+        if parse_dates and not index_col:
```
I think this will also give you true when a list with column names is passed? (and in such a case it should not raise this warning). So I think you should specifically test for True (`parse_dates is True`)

jorisvandenbossche commented on pandas/io/tests/test_excel.py:
```diff
@@ -167,9 +167,10 @@ def test_parse_cols_int(self):
+        df1 = self.get_exceldf('test1', 'Sheet1', index_col=0, parse_dates=True,
                                parse_cols=3)
+        df2 = self.get_exceldf('test1', 'Sheet2', skiprows=[1], index_col=0,
+                               parse_dates=True, parse_cols=3)
```
Why did you change these tests?

jorisvandenbossche commented:
@grahamjeffries Thanks for working on this!

I am only not sure why you changed many of the tests by adding `parse_dates=True`. Why is that needed? 
Besides that, I think it would be good to add a test that actually tests the `parse_dates` keyword behaviour.

grahamjeffries commented:
Okay, I'll change the warning so that it is raised only when `parse_dates is True`, add a test, and and update the documentation example.

Re: changes to the tests, in PR #11870 I removed those arguments in the test to avoid raising the `NotImplemented` error which is raised by the `read_excel` modifications in that same PR. Now, I've removed those modifications so I thought I'd add the arguments back. It turns out that it makes no difference to the testing outcomes, so I'm happy to go either way. Please advise which route is preferable.

jorisvandenbossche commented:
For the tests, if it does not make a difference, you don't need to add this back. Rather try to add a test where it *does* make a difference, so we test the functionality explicitly. 
Eg for the test1.xlsx, the first column is formatted as datetime in excel, not as a string, so indeed in those cases `parse_dates=True` is not needed.

jreback commented:
@grahamjeffries can you update

jreback commented:
can you update according to @jorisvandenbossche comments (and rebase)

jorisvandenbossche commented:
@grahamjeffries Do you have time to update this?

jreback commented:
this would be nice for 0.18.0, but not going to hold us up

jreback commented on doc/source/whatsnew/v0.18.0.txt:
```diff
@@ -481,7 +481,7 @@ Bug Fixes
 - Bug in ``Index`` prevents copying name of passed ``Index``, when a new name is not provided (:issue:`11193`)
 - Bug in ``read_excel`` failing to read any non-empty sheets when empty sheets exist and ``sheetname=None`` (:issue:`11711`)
-- Bug in ``read_excel`` failing to raise ``NotImplemented`` error when keywords ``parse_dates`` and ``date_parser`` are provided (:issue:`11544`)
+- Bug in ``read_excel`` failing to raise warning when keyword ``parse_dates`` and is provided without keyword ``index_col`` (:issue:`11544`)
```
move to 0.18.1

jreback commented:
@grahamjeffries can you rebase / update

jorisvandenbossche commented:
@grahamjeffries Do you have some time to update/rebase this?

jreback commented:
status of this?

needs a rebase

jorisvandenbossche commented:
Either this should be merged, or I should fix the docs that use this keyword (now error during doc build), so I would keep it at the 0.19.0 milestone for now to remember that

jreback commented:
can you rebase / update?

jreback commented:
@grahamjeffries can you rebase / update this?

jorisvandenbossche commented:
I rebased and updated this in https://github.com/pydata/pandas/pull/14326
